server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -913,37 +914,63 @@ private List<String> cleanupStaleRootFiles(Collection<SnapshotId> deletedSnapsho
         return Collections.emptyList();
     }
 
-    private long cleanupStaleIndices(Map<String, BlobContainer> foundIndices, Set<String> survivingIndexIds) {
-        long deleteResult = 0;
+    private void cleanupStaleIndices(Map<String, BlobContainer> foundIndices,
+                                     Set<String> survivingIndexIds,
+                                     ActionListener<Long> listener) {
+        final ActionListener<Long> groupedListener = new MultiActionListener<>(
+            foundIndices.size() - survivingIndexIds.size(), Collectors.summingLong(Long::longValue), listener);
         try {
+            final BlockingQueue<Map.Entry<String, BlobContainer>> staleIndicesToDelete = new LinkedBlockingQueue<>();
             for (Map.Entry<String, BlobContainer> indexEntry : foundIndices.entrySet()) {
-                final String indexSnId = indexEntry.getKey();
-                try {
-                    if (survivingIndexIds.contains(indexSnId) == false) {
-                        LOGGER.debug("[{}] Found stale index [{}]. Cleaning it up", metadata.name(), indexSnId);
-                        indexEntry.getValue().delete();
-                        //This is not the exact number of blobs, as the blob container could also have children.
-                        //However fetching the blobs before deletion is an expensive operation.
-                        //This is ok because currently the deletes are not used. BlobContainer.delete()
-                        //could return the number of blobs and the bytes deleted if we plan to use this
-                        //data somewhere.
-                        deleteResult++;
-                        LOGGER.debug("[{}] Cleaned up stale index [{}]", metadata.name(), indexSnId);
-                    }
-                } catch (Exception e) {
-                    LOGGER.warn(() -> new ParameterizedMessage(
-                        "[{}] index {} is no longer part of any snapshots in the repository, " +
-                        "but failed to clean up their index folders", metadata.name(), indexSnId), e);
+                if (survivingIndexIds.contains(indexEntry.getKey()) == false) {
+                    staleIndicesToDelete.put(indexEntry);
                 }
             }
+
+            // Start as many workers as fit into the snapshot pool at once at the most
+            final int workers = Math.min(
+                threadPool.stats().stream()
+                    .filter(stats -> ThreadPool.Names.SNAPSHOT.equals(stats.name()))
+                    .map(ThreadPoolStats.Stats::active)
```

Comment:
I was not sure why the original code used `max`.

Reply:
I think the number of free snapshot threads can be obtained by `threads - active`. But actually I should refactor? to not explicitly control the number of threads.